hazelcast/src/main/java/com/hazelcast/sql/impl/type/converter/BooleanConverter.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.type.converter;
+
+import com.hazelcast.sql.impl.type.QueryDataTypeFamily;
+
+/**
+ * Converter for {@link java.lang.Boolean} type.
+ */
+public final class BooleanConverter extends Converter {
```

Comment:
The design document specifies that `BIT` should be convertible to `VARCHAR`, `BIT`, `TINYINT`, `SMALLINT`, `INT`, `BINGINT`, `DECIMAL`, `REAL`  and `DOUBLE`. But the converter supports only conversions to `VARCHAR` and `BIT`. Could you please check other converters as well.

Also, it's not clear how converters in general distinguish between implicit and explicit conversion: AFAIU, currently if a converter supports some kind of a conversion it would be applied no matter is it implicit or explicit. For instance, `... WHERE LENGTH(intColumn) = 1` is considered valid in the current code base while the design document treats `INT` to `VARCHAR` conversion as explicit. Looks like we need to convey that implicit-explicit information somehow from the converters or by some other means. Another set of methods? Like `canConvertImplicitly/Explicitly` and probably `implicit/explicitAs*`.

Reply:
There are two ways how BIT could be managed:
1. SQL standard and some vendors restrict conversions between BIT and numeric types
1. But some vendors, e.g. MySQL, treat `BIT` as another numeric type
In my initial implementation, I followed MySQL approach. Then I decided to follow the ANSI standard, as it will simplify the implementation of expressions, but forgot to update the compatibility matrix. It is fixed now.

Regarding implicit/explicit, this is not a converter's concern. Instead, conversion rules are managed by expressions. We may think of any function as a family of overridden functions with different argument types. And during plan preparation, we resolve the correct instance of the function. But resolution rules are typically more complex than "can I convert this to that implicitly?". My previous example demonstrates this: 
1. `INT(11) + BIGINT(20)` is resolved as `CAST(INT(11) AS BIGINT(20) + BIGINT(20))`
1. `INT(11) + VARCHAR` is resolved as `CAST(INT(11) AS DECIMAL) + CAST(VARCHAR as DECIMAL)`
1. `DATE + INT(11)` is failed; not because of an invalid implicit cast for `INT`, but because there is only one valid override for `DATE`, which is `DATE + INTERVAL`, and `INT` cannot be converted to `INTERVAL`, neither implicitly, nor explicitly. 

As a result, there is no immediate need for implicit/explicit checks right now. Maybe we will come with some uniform abstraction for implicit/explicit cast checks, but it is unlikely to be a converter's concern anyway.

Reply:
I see, so the converters basically define the set of valid type pairs for `CAST` operator, which are just the set of valid explicit conversions. One thing that I can't understand, what is the purpose of `precedence/withHigherPrecedence` then? Looks like it just prioritizes implicit conversion for one type over the other for a pair of values of certain types, but doesn't care about the implicit conversion possibility (which is defined by the type conversion matrix in the design document, BTW).

Reply:
Converters define the logic of conversion between type, irrespectively of conversion type (implicit or explicit). Precedence determines which function override should be called for the given argument(s). Every function has own rules, but as general guidance, explicit conversions should not be performed without `CAST` operator. I.e. the implementation of expression should throw an exception in case the conversion is required for the given function override, but it is not implicit. 
So that table should be considered as a guide for expression implementors.

Reply:
And `withHigherPrecedence` is a kind of helper method, which helps expressions to pick the right function override. But this is not the strict rule, only a helper method, which defines part of the selection logic. The example with `+` function and `VARCHAR` demonstrates this well: `VARCHAR + INT` is resolved not as "pick the type with higher precedence", but instead as:
1. If there is a `VARCHAR`, then convert it to `DECIMAL`
2. Then pick the type with higher precedence